File: pt5/1162.as-far-from-land-as-possible.go
package main

func maxDistance(grid [][]int) int {
	n := len(grid)
	type point struct {
		row int
		col int
	}
	queue := []point{}

	for r := 0; r < n; r++ {
		for c := 0; c < n; c++ {
			if grid[r][c] == 1 {
				queue = append(queue, point{row: r, col: c})
			}
		}
	}

	if len(queue) == 0 || len(queue) == n*n {
		return -1
	}

	directions := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	distance := -1
	for len(queue) > 0 {
		distance++
		levelSize := len(queue)
		for i := 0; i < levelSize; i++ {
			current := queue[0]
			queue = queue[1:]
			for _, direction := range directions {
				nextRow := current.row + direction[0]
				nextCol := current.col + direction[1]
				if nextRow >= 0 && nextRow < n && nextCol >= 0 && nextCol < n && grid[nextRow][nextCol] == 0 {
					grid[nextRow][nextCol] = 1
					queue = append(queue, point{row: nextRow, col: nextCol})
				}
			}
		}
	}

	return distance
}

File: pt5/1162.as-far-from-land-as-possible.java
import java.util.*;

class Solution {
    public int maxDistance(int[][] grid) {
        int n = grid.length;
        Queue<int[]> queue = new ArrayDeque<>();

        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (grid[r][c] == 1) {
                    queue.offer(new int[] {r, c});
                }
            }
        }

        if (queue.isEmpty() || queue.size() == n * n) {
            return -1;
        }

        int distance = -1;
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while (!queue.isEmpty()) {
            distance++;
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                int[] cell = queue.poll();
                for (int[] direction : directions) {
                    int nr = cell[0] + direction[0];
                    int nc = cell[1] + direction[1];
                    if (nr >= 0 && nr < n && nc >= 0 && nc < n && grid[nr][nc] == 0) {
                        grid[nr][nc] = 1;
                        queue.offer(new int[] {nr, nc});
                    }
                }
            }
        }

        return distance;
    }
}

